src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineStacksPanel.kt

```diff
@@ -1,64 +1,360 @@
+        override fun paused(suspendContext: SuspendContext) {
+            emptyText.component.isVisible = false
+            buildCoroutineGraph(suspendContext)
+        }
+
+        override fun resumed(suspendContext: SuspendContext?) {
+            panelContent.removeAll()
+        }
+    }
+
+    fun buildCoroutineGraph(suspendContext: SuspendContext) {
+        val suspendContextImpl = suspendContext as? SuspendContextImpl ?: run {
+            emptyText.text = CoroutineStacksBundle.message("coroutine.stacks.could.not.be.built")
+            return
+        }
+
+        val coroutineInfoCache = CoroutineDebugProbesProxy(suspendContextImpl).dumpCoroutines()
+        val coroutineInfoDataList = coroutineInfoCache.cache
+        val dispatcherToCoroutineDataList = mutableMapOf<String, MutableList<CoroutineInfoData>>()
+        for (info in coroutineInfoDataList) {
+            val dispatcher = info.descriptor.dispatcher ?: continue
+            dispatcherToCoroutineDataList.computeIfAbsent(dispatcher) { mutableListOf() }.add(info)
+        }
+
+        val dispatcherLabel = JLabel(CoroutineStacksBundle.message("select.dispatcher"))
+        val coroutineStacksWindowHeader = Box.createHorizontalBox()
+        val dispatchers = dispatcherToCoroutineDataList.keys.toTypedArray()
+        val dispatcherDropdownMenu = ComboBox(dispatchers)
+
+        dispatcherDropdownMenu.addActionListener {
+            val selectedDispatcher = dispatcherDropdownMenu.selectedItem as String
+            updateCoroutineStackForest(selectedDispatcher, dispatcherToCoroutineDataList, suspendContextImpl)
+        }
+
+        val comboBoxSize = Dimension(Constants.comboBoxHeight, Constants.comboBoxWidth)
+        dispatcherDropdownMenu.apply {
+            preferredSize = comboBoxSize
+            maximumSize = comboBoxSize
+            minimumSize = comboBoxSize
+        }
+        coroutineStacksWindowHeader.add(dispatcherLabel)
+        coroutineStacksWindowHeader.add(dispatcherDropdownMenu)
+
+        val selectedDispatcher = dispatcherDropdownMenu.selectedItem as String
+        updateCoroutineStackForest(selectedDispatcher, dispatcherToCoroutineDataList, suspendContextImpl)
+        panelContent.add(coroutineStacksWindowHeader)
+        panelContent.add(forest)
+        add(panelContent)
+    }
+
+    private fun updateCoroutineStackForest(
+        selectedDispatcher: String,
+        dispatcherToCoroutineDataList: MutableMap<String, MutableList<CoroutineInfoData>>,
+        suspendContextImpl: SuspendContextImpl
+    ) {
+        forest.removeAll()
+        val root = Node(null, -1, mutableMapOf(), "")
+        val coroutineStackForest = buildCoroutineStackForest(
+            root,
+            dispatcherToCoroutineDataList[selectedDispatcher]!!,
+            suspendContextImpl
+        )
+        forest.add(coroutineStackForest)
+        updateUI()
+    }
+
+    private fun buildCoroutineStackForest(
```

Comment:
Speaking about this function, it is very hard to read it. We have 3 stages in the forest building algorithm:
1. Stack frame graph building
2. Creating a forest of coroutine traces
3. Creating UI

However, I can't see these stages while reading this function, which makes it extremely hard to understand and maintain. I made a sketch of how it should ideally look like:
```
    data class CoroutineTrace(val locations: MutableList<Location>, val header: String, val hoverContent: String)
    
    private fun SuspendContextImpl.buildCoroutineStackForest(
        rootValue: Node,
        coroutineDataList: MutableList<CoroutineInfoData>,
    ): JBScrollPane {
        val lastRunningStackFrame = buildStackFrameGraph(coroutineDataList, rootValue)
        val coroutineTraces = createCoroutineTraces(rootValue)
        return createCoroutineTraceForest(coroutineTraces, lastRunningStackFrame)
    }

    private fun SuspendContextImpl.createCoroutineTraceForest(traces: List<CoroutineTrace?>, lastRunningStackFrame: String): JBScrollPane {
        val componentData: MutableList<Component> = mutableListOf()
        var previousListSelection: JBList<*>? = null
        traces.forEach { trace ->
            if (trace == null) {
                componentData.add(Separator())
                return@forEach
            }
            val vertex = createCoroutineTraceUI(trace, lastRunningStackFrame)
            vertex.addListSelectionListener { e ->
                val currentList = e.source as? JBList<*> ?: return@addListSelectionListener
                if (previousListSelection != currentList) {
                    previousListSelection?.clearSelection()
                }
                previousListSelection = currentList
            }
            componentData.add(vertex)
        }

        val forest = ContainerWithEdges()
        componentData.forEach { forest.add(it) }
        forest.layout = ForestLayout()
        return JBScrollPane(forest)
    }

    private fun createCoroutineTraces(rootValue: Node): List<CoroutineTrace?> {
        val stack = Stack<Pair<Node, Int>>()
        val parentStack = Stack<Node>()

        stack.push(Pair(rootValue, 0))
        val locationsFollowedBySeparators = mutableListOf<Location?>()

        var currentTrace: CoroutineTrace? = null
        val coroutineTraces = mutableListOf<CoroutineTrace?>()
        while (stack.isNotEmpty()) {
            val (currentNode, level) = stack.pop()
            val parent = if (parentStack.isNotEmpty()) parentStack.pop() else null

            if (parent != null) {
                if (parent.num != currentNode.num) {
                    currentTrace = CoroutineTrace(
                        mutableListOf(currentNode.stackFrameLocation),
                        "${currentNode.num} Coroutines ACTIVE",
                        currentNode.coroutinesActive
                    )
                    coroutineTraces.add(currentTrace)
                } else {
                    currentTrace?.locations?.add(currentNode.stackFrameLocation)
                }
            }

            currentNode.children.values.reversed().forEachIndexed { index, child ->
                if (index != currentNode.children.values.size - 1) {
                    locationsFollowedBySeparators.add(child.stackFrameLocation)
                    coroutineTraces.add(null)
                }
                stack.push(child to (level + 1))
                parentStack.push(currentNode)
            }
        }

        return coroutineTraces
    }

    private fun SuspendContextImpl.createCoroutineTraceUI(trace: CoroutineTrace, lastRunningStackFrame: String): JBList<String> {
        val vertex = JBList<String>()
        val data = mutableListOf<String>()
        data.add(trace.header)
        data.addAll(trace.locations.map { it.toString() })
        vertex.setListData(data.toTypedArray())
        val lastStackFrame = data[trace.locations.size]
        vertex.border = if (lastRunningStackFrame == lastStackFrame) {
            BorderFactory.createLineBorder(JBColor.GRAY)
        } else {
            BorderFactory.createLineBorder(JBColor.BLACK)
        }
        vertex.cellRenderer = CustomCellRenderer(trace.hoverContent)
        vertex.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent?) {
                val list = e?.source as? JBList<*> ?: return
                val index = list.locationToIndex(e.point).takeIf { it > 0 } ?: return
                val location = trace.locations[index - 1]
                debugProcess.managerThread.invoke(object : DebuggerCommandImpl() {
                    override fun action() {
                        val positionManager = debugProcess.positionManager
                        val sourcePosition = positionManager.getSourcePosition(location)
                        sourcePosition?.navigate(true)
                    }
                })
            }
        })

        return vertex
    }

```

So now the `buildCoroutineStackForest` function looks like this:
```
private fun SuspendContextImpl.buildCoroutineStackForest(
        rootValue: Node,
        coroutineDataList: MutableList<CoroutineInfoData>,
    ): JBScrollPane {
        val lastRunningStackFrame = buildStackFrameGraph(coroutineDataList, rootValue)
        val coroutineTraces = createCoroutineTraces(rootValue)
        return createCoroutineTraceForest(coroutineTraces, lastRunningStackFrame)
    }
```
It is just 3 lines long, but we can clearly see what is going on.